crates/package_manager/src/add.rs

```diff
@@ -0,0 +1,96 @@
+use crate::Install;
+use derive_more::{Display, Error};
+use miette::Diagnostic;
+use pacquet_lockfile::Lockfile;
+use pacquet_npmrc::Npmrc;
+use pacquet_package_json::PackageJsonError;
+use pacquet_package_json::{DependencyGroup, PackageJson};
+use pacquet_registry::{PackageTag, PackageVersion};
+use pacquet_tarball::Cache;
+use reqwest::Client;
+
+/// This subroutine does everything `pacquet add` is supposed to do.
+#[must_use]
+pub struct Add<'a, ListDependencyGroups, DependencyGroupList>
+where
+    ListDependencyGroups: Fn() -> DependencyGroupList,
+    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
+{
+    /// Shared cache that store downloaded tarballs.
+    pub tarball_cache: &'a Cache,
+    /// HTTP client to make HTTP requests.
+    pub http_client: &'a Client,
+    /// Configuration read from `.npmrc`.
+    pub config: &'static Npmrc,
+    /// Data from the `package.json` file.
+    pub package_json: &'a mut PackageJson,
+    /// Data from the `pnpm-lock.yaml` file.
+    pub lockfile: Option<&'a Lockfile>,
+    /// List of [`DependencyGroup`]s.
+    pub list_dependency_groups: ListDependencyGroups,
+    /// Name of the package to add.
+    pub package: &'a str,
+    /// Whether `--save-exact` is provided.
+    pub save_exact: bool,
+}
+
+/// Error type of [`Add`].
+#[derive(Debug, Display, Error, Diagnostic)]
+pub enum AddError {
+    #[display("Failed to add package to manifest: {_0}")]
+    AddDependencyToPackageJson(#[error(source)] PackageJsonError),
+    #[display("Failed save the manifest file: {_0}")]
+    SavePackageJson(#[error(source)] PackageJsonError),
+}
+
+impl<'a, ListDependencyGroups, DependencyGroupList>
+    Add<'a, ListDependencyGroups, DependencyGroupList>
+where
+    ListDependencyGroups: Fn() -> DependencyGroupList,
+    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
+{
+    pub async fn run(self) -> Result<(), AddError> {
+        let Add {
+            tarball_cache,
+            http_client,
+            config,
+            package_json,
+            lockfile,
+            list_dependency_groups,
+            package,
+            save_exact,
+        } = self;
+
+        let latest_version = PackageVersion::fetch_from_registry(
+            package,
+            PackageTag::Latest, // TODO: add support for specifying tags
+            http_client,
+            &config.registry,
+        )
+        .await
+        .expect("resolve latest tag"); // TODO: properly propagate this error
+
+        let version_range = latest_version.serialize(save_exact);
+        for dependency_group in list_dependency_groups() {
+            package_json
+                .add_dependency(package, &version_range, dependency_group)
+                .map_err(AddError::AddDependencyToPackageJson)?;
+        }
+
+        Install {
+            tarball_cache,
+            http_client,
+            config,
+            package_json,
+            lockfile,
+            dependency_groups: list_dependency_groups(),
```

Comment:
what is the benefit of passing the dependency_groups to the add function via an anonymous function? Why not just passing dependency_groups directly?

Reply:
Unlike in other subroutine, this `dependecy_groups` in `Add` is used twice (with move semantics).

Reply:
Wouldn't it make more sense to copy it, instead of potentially recalculating it several time, when running a function?

Reply:
The function is `AddArgs:: dependency_groups`. It's very cheap (it's just a few if statements). Allocating a `Vec` is more expensive.